### ISSUE TITLE
AOTAutograd Tracing

### ISSUE THREAD

**Issue by jeromeku** [thuml/depyf#77]
@youkaichao 

Thanks for the great investigation into the inner workings of `torch.compile`.

Hoping you could shed light on the following issue around how AOTAutograd decomposes high level `torch IR / ops` -> `aten IR` in the context of tensor subclass tracing and pre / post grad `inductor` passes:

Concretely, take the decomposition of `torch.nn.functional.linear` as an example.

```python
import torch
from torch._inductor.decomposition import select_decomp_table
from torch.fx import GraphModule
from torch.fx.experimental.proxy_tensor import make_fx

device = "cuda"
x = torch.randn(10, 10).to(device)
y = torch.randn(10, 10).to(device)

g: GraphModule = make_fx(torch.nn.functional.linear, select_decomp_table(), tracing_mode="real")(x, y)
g.print_readable()
```
This prints
```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[10, 10]", arg1_1: "f32[10, 10]"):
        # No stacktrace found for following nodes
        permute: "f32[10, 10]" = torch.ops.aten.permute.default(arg1_1, [1, 0]);  arg1_1 = None
        mm: "f32[10, 10]" = torch.ops.aten.mm.default(arg0_1, permute);  arg0_1 = permute = None
        return mm
```
I've tried to step through the code using `pdb` but still can't figure out how the decomposition of `linear` is happening:
- where is the mapping of `linear` to `permute` and `mm`?
- if there is no direct mapping, how can I determine where along the trace these registered ops are being called?

Noticed that something similar is called during the pregrad passes that `inductor` runs to normalize the IR into something more amenable for downstream passes.

Also related are the [AOTAutograd](https://github.com/pytorch/workshops/tree/master/ASPLOS_2024/aot_autograd.pdf) slides from ASPLOS 2024.  On slide 23, it is mentioned that AOTAutograd handles "operator decomposition" among other Pytorch framework related functions that occur at the C++ level, and it does this through `__torch_dispatch__`.  

Where / how does this call back to Python happen after the dispatching, autograd, and decomposition work is handled in C++?

Thanks!

-Jerome
